src/pybacked.rs

```diff
@@ -79,13 +88,15 @@ impl FromPyObject<'_> for PyBackedStr {
 /// A wrapper around `[u8]` where the storage is either owned by a Python `bytes` object, or a Rust `Box<[u8]>`.
 ///
 /// This type gives access to the underlying data via a `Deref` implementation.
+#[derive(Clone)]
```

Comment:
Isn't the derived `Clone` impl incorrect for the `Rust` variant? Cloning `Box<[u8]>` will yield a new allocation for `data` will still point to the old one meaning this is a use-after-free in waiting?

So we either need to manually fix up the `data` pointer or use a shared ownership pointer for the Rust side as well, e.g. `Arc<[u8]>`.

Reply:
(And so yes, these do need tests. ;-))

Reply:
Yes I was thinking similarly about the `Box`, an `Arc` seems reasonable to me.

I also half wonder if it was a mistake to have the Rust variant here given the name `PyBackedBytes`. I added it to support bytearrays in the same way we already do for `Cow<[u8]>` but on further reflection it does seem a bit add odds with the name that we do the copy.

I think probably it's ok as I would expect most uses to be with bytes, but we might want to document this copy-from-bytearray more clearly and justify for sake of safety.

Reply:
> Isn't the derived `Clone` impl incorrect for the `Rust` variant? Cloning `Box<[u8]>` will yield a new allocation for `data` will still point to the old one meaning this is a use-after-free in waiting?

Big oops, thanks for catching.

Using `Arc<[u8]>` sounds good to me aswell, that way this stays a cheap clone either way.